### PR TITLE
test: introduce beacon api test ignore list

### PR DESCRIPTION
**Pull request by jeluard**
**Motivation**

Make sure lodestar supports latest [Beacon APIs](https://github.com/ethereum/beacon-APIs/releases). 

**Description**

To allow for easier testing against latest Beacon API, a black-list of not-yet supported features/examples is introduced.

Relates to #5710
Builds on top of https://github.com/ChainSafe/lodestar/compare/unstable...nflaig/update-beacon-spec-version

**Steps to test or reproduce**

All unit tests in `packages/api` should pass.

### CODE REVIEW

File: packages/api/src/beacon/routes/beacon/pool.ts [hunk 1]
```diff
@@ -63,7 +63,7 @@ export type Api = {
    * @returns any Successful response
    * @throws ApiError
    */
-  getPoolBlsToExecutionChanges(): Promise<
+  getPoolBLSToExecutionChanges(): Promise<
```
Comment (nflaig) [inline packages/api/src/beacon/routes/beacon/pool.ts]: It does not make sense to do the method renaming on unstable branch, we can apply these changes to https://github.com/ChainSafe/lodestar/pull/6080, otherwise it's just duplicate work and merge conflicts

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 2]
```diff
@@ -84,11 +84,83 @@ const testDatas = {
   ...validatorTestData,
 };
 
+const ignoredOperationsIds = [
+  /*
+   #5693
+   missing finalized
+   */
+  "getStateRoot",
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: Excluding the whole route because of one missing field is worse than checking against an older version of the spec in my opinion.

Ignore list is only acceptable for routes that are not yet implemented. A generic approach to ignore missing properties per route would be ideal.

File: packages/api/src/beacon/routes/beacon/pool.ts (hunk 1 above)
Comment (jeluard) [reply packages/api/src/beacon/routes/beacon/pool.ts]: Ok I will migrate the changes then.

File: packages/api/test/unit/beacon/oapiSpec.test.ts (hunk 2 above)
Comment (jeluard) [reply packages/api/test/unit/beacon/oapiSpec.test.ts]: >  A generic approach to ignore missing properties per route would be ideal.

I will check that option.

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 3]
```diff
@@ -84,11 +84,90 @@ const testDatas = {
   ...validatorTestData,
 };
 
+const ignoredOperations = [
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts:87]: Just looking at this, it is not immediately clear how it works. Maybe it makes sense to have separate ignore lists
- ignoredRoutes
- ignoredReqParams
- ignoredResParams

or similar

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 4]
```diff
@@ -84,11 +84,90 @@ const testDatas = {
...
+   /query/syncing_status - must be integer
+   */
+  {id: "getHealth"},
+
+  /* 
+   #4638 
+   /query - must have required property 'skip_randao_verification'
+   */
+  {id: "produceBlockV2"},
+  {id: "produceBlindedBlock"},
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: It would be good if we don't have to skip the whole route just because of one invalid / missing query param

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 5]
```diff
@@ -19,17 +19,40 @@ ajv.addKeyword({
...
+
+/**
+ * Run all tests defined as part of the OpenAPI spec provided as `openApiJson`.
+ *
+ * @param openApiJson
+ * @param routesData
+ * @param reqSerializers
+ * @param returnTypes
+ * @param testDatas
+ * @param opts
```
Comment (nflaig) [inline packages/api/test/utils/checkAgainstSpec.ts]: jsdoc like this is not very informative, the only param that is documented here is `filteredOperations`, alternatively could create a type for that and add jsdoc there

File: packages/api/test/unit/beacon/oapiSpec.test.ts (hunk 4 above)
Comment (jeluard) [reply packages/api/test/unit/beacon/oapiSpec.test.ts]: Good point, will do something similar as what I did for the required response properties.

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 5 above)
Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts]: Makes sense.

File: packages/api/test/unit/beacon/oapiSpec.test.ts (hunk 3 above)
Comment (jeluard) [reply packages/api/test/unit/beacon/oapiSpec.test.ts:87]: I will clarify that. As those are per operation (and not global), it felt more natural to have some operation level extra data.

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 6]
```diff
@@ -5,31 +5,66 @@ import {applyRecursively, OpenApiJson, parseOpenApiSpec, ParseOpenApiSpecOpts} f
...
+  /**
+   * When specified, associated operation is tested but `required` properties are ignored from response.
+   */
+  requiredResponseProperties?: string[];
+};
+
+/**
+ * @returns `true` if the `filteredOperation` is defined and no required property is specified.
+ */
+function isOperationIgnored(filteredOperation: FilteredOperation | undefined): boolean {
```
Comment (nflaig) [inline packages/api/test/utils/checkAgainstSpec.ts]: wouldn't it be better to just add a boolean like `ignoreRoute` or `ignoreOperation`? This would make it more explicit

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 7]
```diff
@@ -5,31 +5,66 @@ import {applyRecursively, OpenApiJson, parseOpenApiSpec, ParseOpenApiSpecOpts} f
 import {GenericServerTestCases} from "./genericServerTest.js";
 
 const ajv = new Ajv({
-  // strict: true,
-  // strictSchema: true,
```
Comment (nflaig) [inline packages/api/test/utils/checkAgainstSpec.ts:9]: Why are these lines removed and what are the implications?

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 8]
```diff
@@ -5,31 +5,66 @@ import {applyRecursively, OpenApiJson, parseOpenApiSpec, ParseOpenApiSpecOpts} f
...
   allErrors: true,
 });
 
-// TODO: Still necessary?
+// Ensure embedded schema 'example' do not fail validation
```
Comment (nflaig) [inline packages/api/test/utils/checkAgainstSpec.ts:13]: Did you check if example keyword here can even be removed? i remember this is a very old issue, maybe ajv fixed this on their end by now

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 9]
```diff
@@ -84,11 +84,102 @@ const testDatas = {
   ...validatorTestData,
 };
 
+const filteredOperations: FilteredOperation[] = [
+  /*
+   #5693
+   missing finalized
+   */
+  {id: "getStateRoot", requiredResponseProperties: ["finalized"]},
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: This reads wrong to me, what about

```suggestion
  {id: "getStateRoot", ignoredResponseProperties: ["finalized"]},
```

could also have a single array with "ignore objects", e.g.

```suggestion
  {id: "getStateRoot", ignoredProperties: [{response: "finalized"}]},
```

but this probably makes things more complicated

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 8 above)
Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts:13]: Schemas from the beacon chain embed `example` sections. Those are not valid json schema AFAIK and fail validation.
This essentially ignores those properties.

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 6 above)
Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts]: This would require either adding `ignoreOperation` for all operations with no property filtering, which felt a bit verbose to me. 
Also the whole purpose of being in this list is to be filtered, so felt redundant.

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 7 above)
Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts:9]: No change in behavior are those are comments.
I will check if `strict: true` can be used instead.

Comment (nflaig) [reply packages/api/test/utils/checkAgainstSpec.ts:9]: ah right those were comments, I would assume these are default settings

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 8 above)
Comment (nflaig) [reply packages/api/test/utils/checkAgainstSpec.ts:13]: My comment was specifically regarding if we still need to add `example` as keyword, this is related to previous TODO comment (`TODO: Still necessary?`) because I remember that at some point ajv removed `example` from default keywords. See related issue https://github.com/loopbackio/loopback-next/issues/9308, the workaround is exactly what we did here to add the keyword manually.

> This essentially ignores those properties.

I'd assume the examples should be valid, if not seems an issue with the spec?


Or is this related to removing this 👇 
```
validate: () => true,
errors: false,
```

Was is not validated before but we validate examples now?

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 6 above)
Comment (nflaig) [reply packages/api/test/utils/checkAgainstSpec.ts]: > This would require either adding ignoreOperation for all operations with no property filtering

Can't this just be optional property? We would use the id to know which route to filter and then look at other values to determine what filtering / ignore rule to apply.

Right now it is not self explanatory if you just look at `filteredOperations` array what is happening, having the jsdoc to explain how it works is good but shouldn't be required to understand the code (if possible)

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 8 above)
Comment (nflaig) [reply packages/api/test/utils/checkAgainstSpec.ts:13]: > I'd assume the examples should be valid, if not seems an issue with the spec?

You can ignore this question, I remember now that as per openapi spec examples are allowed but ajv does not have them, that's why we need custom keyword.

I would assume that by default it does not validate if this still works after removing
```
validate: () => true,
errors: false,
```

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 7 above)
Comment (nflaig) [reply packages/api/test/utils/checkAgainstSpec.ts:9]: If validation still passed why not, see https://ajv.js.org/options.html#strict. Although our job here is not to validate the beacon spec schema but rather if our implementation is according to spec. The beacon spec should be a valid schema definition I'd assume

Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts:9]: Actually it almost pass, modulo a small typo in the `Beacon-APIs` repo (see https://github.com/ethereum/beacon-APIs/pull/391)

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 8 above)
Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts:13]: It doesn't validate the content of `example` but allows this property to exist (w/o this call validation fails with e.g. `request schema - strict mode: unknown keyword: "example"`).

> per openapi spec examples 

Ah right, i's valid `openapi` but not `JSON Schema`. As in our tests the whole openapi `schema` is used for `JSON Schema` validation, it fails because of this extra property.
Probably the better solution is to remove those examples from `schema` before validation.
Current solution allows `example` to be anywhere, might be problematic.

Comment (nflaig) [reply packages/api/test/utils/checkAgainstSpec.ts:13]: > Probably the better solution is to remove those examples from schema before validation.
Current solution allows example to be anywhere, might be problematic.

Can't we just keep the previous behavior so that it is considered always valid (`validate: () => true`)?

Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts:13]: 👍

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 10]
```diff
@@ -84,11 +84,105 @@ const testDatas = {
...
+  "getLightClientFinalityUpdate",
+  "getLightClientOptimisticUpdate",
+  "getPoolBLSToExecutionChanges",
+  "submitPoolBLSToExecutionChange",
+
+  /* 
+   #6168
+   /query/syncing_status - must be integer
+   */
+  "getHealth",
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: Isn't it sufficient to just ignore syncing_status query param?

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 11]
```diff
@@ -19,17 +19,35 @@ ajv.addKeyword({
 
 ajv.addFormat("hex", /^0x[a-fA-F0-9]+$/);
 
+/**
+ * A property that will be filtered during tests execution.
```
Comment (nflaig) [inline packages/api/test/utils/checkAgainstSpec.ts]: Could more context here on why this is useful / needed (what you already wrote in the PR description)

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 12]
```diff
@@ -84,11 +84,105 @@ const testDatas = {
   ...validatorTestData,
 };
 
+const filteredOperations = [
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: Here we use term "filtered" but below we use "ignored" (also have ignoredTopics). For me "ignored" makes it a bit clearer

File: packages/api/test/unit/beacon/oapiSpec.test.ts (hunk 10 above)
Comment (jeluard) [reply packages/api/test/unit/beacon/oapiSpec.test.ts]: Good idea, fixed

File: packages/api/test/unit/beacon/oapiSpec.test.ts (hunk 12 above)
Comment (jeluard) [reply packages/api/test/unit/beacon/oapiSpec.test.ts]: Makes sense

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 11 above)
Comment (jeluard) [reply packages/api/test/utils/checkAgainstSpec.ts]: Done

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 13]
```diff
@@ -84,11 +84,105 @@ const testDatas = {
...
-  routesDropOneOf: ["produceBlockV2", "produceBlindedBlock", "publishBlindedBlock"],
-});
+runTestCheckAgainstSpec(
+  openApiJson,
+  routesData,
+  reqSerializers,
+  returnTypes,
+  testDatas,
+  {
+    // TODO: Investigate why schema validation fails otherwise
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: Might be good ref the issue you created for this (to get more context)
- https://github.com/ChainSafe/lodestar/issues/6187

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 14]
```diff
@@ -84,11 +84,105 @@ const testDatas = {
...
+  "getPoolBLSToExecutionChanges",
+  "submitPoolBLSToExecutionChange",
+];
+
+const ignoredProperties: Record<string, IgnoredProperty> = {
+  /*
+   #5693
+   missing finalized
+   */
+  getStateRoot: {responseProperties: ["finalized"]},
```
Comment (nflaig) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: I like the nested property approach, we anyways don't have type safety for the props so string dot notation is perfectly fine I think and semantics are nice

You could even go a step further if you wann to get rid of the objects here altogether

```suggestion
  getStateRoot: ["response.finalized"],
```

or I guess to make it less verbose could also just use "response" as properties is already implied here

```suggestion
  getStateRoot: {response: ["finalized"]},
```

I guess with 2nd approach (currently used) we get type safety on request / response properties which is kinda good

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 15]
```diff
@@ -19,17 +19,69 @@ ajv.addKeyword({
 
 ajv.addFormat("hex", /^0x[a-fA-F0-9]+$/);
 
+/**
+ * A set of properties that will be ignored during tests execution.
+ * This allows for a black-list mechanism to have a test pass while some part of the spec is not yet implemented.
+ *
+ * Properties can be nested using dot notation, follwong JSONPath semantic.
```
Comment (nflaig) [inline packages/api/test/utils/checkAgainstSpec.ts]:
```suggestion
 * Properties can be nested using dot notation, following JSONPath semantic.
```

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 16]
```diff
@@ -19,17 +19,69 @@ ajv.addKeyword({
...
+   * Properties to ignore in the response schema
+   */
+  responseProperties?: string[];
+};
+
+/**
+ * Recursively remove a property from a schema
+ *
+ * @param schema a schema to remove a property from
+ * @param property a JSONPath like property to remove from the schema
```
Comment (nflaig) [inline packages/api/test/utils/checkAgainstSpec.ts]:
```suggestion
 * @param schema Schema to remove a property from
 * @param property JSONPath like property to remove from the schema
```
or
```suggestion
 * @param schema A schema to remove a property from
 * @param property A JSONPath like property to remove from the schema
```

**Review by nflaig — COMMENTED** [review]
Looks pretty good, I like the semantics. Just some minor nits

File: packages/api/test/unit/beacon/oapiSpec.test.ts (hunk 14 above)
Comment (jeluard) [reply packages/api/test/unit/beacon/oapiSpec.test.ts]: 2nd option looks good!

**Review by nflaig — DISMISSED** [review]

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 17]
```diff
@@ -1,16 +1,16 @@
...
 import {ReqGeneric, ReqSerializer, ReturnTypes, RouteDef} from "../../src/utils/types.js";
-import {applyRecursively, OpenApiJson, parseOpenApiSpec, ParseOpenApiSpecOpts} from "./parseOpenApiSpec.js";
+import {applyRecursively, JsonSchema, OpenApiJson, parseOpenApiSpec, ParseOpenApiSpecOpts} from "./parseOpenApiSpec.js";
 import {GenericServerTestCases} from "./genericServerTest.js";
 
 const ajv = new Ajv({
-  // strict: true,
-  // strictSchema: true,
+  strict: true,
+  strictTypes: false, // TODO Enable once https://github.com/ethereum/beacon-APIs/pull/391 is released
```
Comment (nazarhussain) [inline packages/api/test/utils/checkAgainstSpec.ts]: @jeluard Please create an issue and assign to yourself so we don't forget to track it.

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 18]
```diff
@@ -19,17 +19,69 @@ ajv.addKeyword({
...
+ * Recursively remove a property from a schema
+ *
+ * @param schema Schema to remove a property from
+ * @param property JSONPath like property to remove from the schema
+ */
+function deleteNested(schema: JsonSchema | undefined, property: string): void {
+  const properties = schema?.properties;
+  if (property.includes(".")) {
+    // Extract first segment, keep the rest as dotted
+    const [key, rest] = property.split(".", 2);
```
Comment (nazarhussain) [inline packages/api/test/utils/checkAgainstSpec.ts]: Passing `limit` to `split` will not keep the `rest` as dotted. It will `exclude` the remaining values. 

From docs:
> Any leftover text is not included in the array at all.


```js
> "alpha.beta.gamma".split(".", 2)
[ 'alpha', 'beta' ]
```

File: packages/api/test/utils/checkAgainstSpec.ts [hunk 19]
```diff
@@ -68,7 +120,13 @@ export function runTestCheckAgainstSpec(
           stringifyProperties((reqJson as ReqGeneric).params ?? {});
           stringifyProperties((reqJson as ReqGeneric).query ?? {});
 
-          // Validate response
+          const ignoredProperties = ignoredProperty?.request;
+          if (ignoredProperties) {
+            // Remove ignored properties from schema validation
+            ignoredProperties.forEach((property) => deleteNested(routeSpec.requestSchema, property));
```
Comment (nazarhussain) [inline packages/api/test/utils/checkAgainstSpec.ts]: We usually prefer to use loop` instead of `forEach` for performance reason.

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 20]
```diff
@@ -84,11 +84,105 @@ const testDatas = {
...
+  testDatas,
+  {
+    // TODO: Investigate why schema validation fails otherwise (see #6187)
+    routesDropOneOf: [
+      "produceBlockV2",
+      "produceBlockV3",
+      "produceBlindedBlock",
+      "publishBlindedBlock",
+      "publishBlindedBlockV2",
+    ],
```
Comment (nazarhussain) [inline packages/api/test/unit/beacon/oapiSpec.test.ts:162]: Does it still fails even including `ignored properties`?

File: packages/api/test/unit/beacon/oapiSpec.test.ts [hunk 21]
```diff
@@ -84,11 +84,105 @@ const testDatas = {
...
+  getHealth: {request: ["query.syncing_status"]},
+
+  /**
+   * #6185
+   *  - must have required property 'query'
+   */
+  getBlobSidecars: {request: ["query"]},
+
+  /* 
+   #4638
```
Comment (nazarhussain) [inline packages/api/test/unit/beacon/oapiSpec.test.ts]: It's your choice, but I prefer to have a complete link eg. `https://github.com/ChainSafe/lodestar/issues/4638` in comments, easier to identify if we are pointing to PR or issue and also easy to navigate with one click.

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: packages/api/test/utils/checkAgainstSpec.ts (hunk 18 above)
Comment (nflaig) [reply packages/api/test/utils/checkAgainstSpec.ts]: Good catch, I guess in practice this does not happen as we only have one level of nesting. Maybe just remove recursively calling deleteNested?

or could do
```js
const [key, ...rest] = property.split(".");
deleteNested(properties?.[key], rest.join("."));
```

File: packages/api/test/unit/beacon/oapiSpec.test.ts (hunk 20 above)
Comment (nflaig) [reply packages/api/test/unit/beacon/oapiSpec.test.ts:162]: This a different issue, unrelated to properties
- see https://github.com/ChainSafe/lodestar/issues/6187

Ignoring the whole route would solve it but this is better

**Comment by jeluard** [comment]
@nazarhussain your points have been addressed

**Review by nazarhussain — APPROVED** [review]

**Review by nflaig — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]
lgtm :+1:

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.14.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.14.0) :tada:
